zk/conn.go

```diff
@@ -266,13 +280,29 @@ func WithEventCallback(cb EventCallback) connOption {
 	}
 }
 
+func (c *Conn) closeWithError(err error) {
+	c.closeErrMu.Lock()
+	c.closeErr = err
+	c.closeErrMu.Unlock()
+	atomic.StoreInt32(&c.closed, 1)
+
+	// The process of flushing requests is racy by design so do it in two steps.
```

Comment:
queueRequest needs to check c.closed. If it's closed, you need to create a response{} with an error, put that into a buffered channel, and return that channel immediately. That eliminates the race, right?